File: cpp/cpp-std/demo15a-deadlock.cpp
/*
DEADLOCK
Version A
*/


#include <iostream>
#include <string>
#include <thread>
#include <mutex>
using namespace std;



std::mutex mut;



void doTask(std::string name) {
    mut.lock();

    cout << name << " acquired resource" << endl;

    // mut.unlock(); // Forget this statement ==> deadlock
}



int main() {
    auto thFoo = std::thread(&doTask, "foo");
    auto thBar = std::thread(&doTask, "bar");

    thFoo.join();
    thBar.join();

    cout << "You will never see this statement due to deadlock!" << endl;
    return 0;
}

File: python/demo15a_deadlock.py
'''
DEADLOCK
Version A
'''

import threading



mutex = threading.Lock()



def do_task(name: str):
    mutex.acquire()
    print(f'{name} acquired resource')
    # mutex.release() # Forget this statement ==> deadlock



th_foo = threading.Thread(target=do_task, args=('foo',))
th_bar = threading.Thread(target=do_task, args=('bar',))

th_foo.start()
th_bar.start()
th_foo.join()
th_bar.join()

print('You will never see this statement due to deadlock!')
